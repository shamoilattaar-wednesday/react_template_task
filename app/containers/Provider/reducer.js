import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: trackTypes, Creators: trackCreators } = createActions({
  requestGetTrackDetails: ['trackName'],
  successGetTrackDetails: ['data'],
  failureGetTrackDetails: ['error'],
  requestGetTrackDetailsById: ['trackId'],
  successGetSingleTrackDetails: ['singledata'],

  clearTrackDetails: {}
});
export const initialState = { trackName: null, tracksData: {}, tracksError: null, trackId: null, singletrackData: {} };

export const trackReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case trackTypes.REQUEST_GET_TRACK_DETAILS:
        draft.trackName = action.trackName;
        break;
      case trackTypes.CLEAR_TRACK_DETAILS:
        draft.trackName = null;
        draft.tracksError = null;
        draft.tracksData = {};
        draft.trackId = null;
        draft.singletrackData = {};
        break;
      case trackTypes.SUCCESS_GET_TRACK_DETAILS:
        draft.tracksData = action.data;
        break;
      case trackTypes.FAILURE_GET_TRACK_DETAILS:
        draft.tracksError = get(action.error, 'message', 'something_went_wrong');
        break;
      case trackTypes.REQUEST_GET_TRACK_DETAILS_BY_ID:
        draft.trackId = action.trackId;
        break;
      case trackTypes.SUCCESS_GET_SINGLE_TRACK_DETAILS:
        draft.singletrackData = action.singledata;
        break;
    }
  });

export default trackReducer;
