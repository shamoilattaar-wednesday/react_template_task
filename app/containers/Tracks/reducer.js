import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: trackTypes, Creators: trackCreators } = createActions({
  requestGetTrackDetails: ['trackName'],
  successGetTrackDetails: ['data'],
  failureGetTrackDetails: ['error'],
  clearTrackDetails: {}
});
export const initialState = { trackName: null, tracksData: {}, tracksError: null };

export const trackReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case trackTypes.REQUEST_GET_TRACK_DETAILS:
        draft.trackName = action.trackName;
        break;
      case trackTypes.CLEAR_TRACK_DETAILS:
        draft.trackName = null;
        draft.tracksError = null;
        draft.tracksData = {};
        break;
      case trackTypes.SUCCESS_GET_TRACK_DETAILS:
        draft.tracksData = action.data;
        break;
      case trackTypes.FAILURE_GET_TRACK_DETAILS:
        draft.tracksError = get(action.error, 'message', 'something_went_wrong');
        break;
    }
  });

export default trackReducer;
