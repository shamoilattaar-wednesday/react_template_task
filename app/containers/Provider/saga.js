import { put, call, takeLatest } from 'redux-saga/effects';
import { getTracks, getTrackById } from '@services/repoApi';
import { trackTypes, trackCreators } from './reducer';

const { REQUEST_GET_TRACK_DETAILS, REQUEST_GET_TRACK_DETAILS_BY_ID } = trackTypes;
const { successGetTrackDetails, failureGetTrackDetails, successGetSingleTrackDetails } = trackCreators;
export function* getTrackDetails(action) {
  const response = yield call(getTracks, action.trackName);
  const { data, status } = response;
  if (status === 200) {
    yield put(successGetTrackDetails(data));
  } else {
    yield put(failureGetTrackDetails(data));
  }
}
export function* getTrackDetailsById(action) {
  const response = yield call(getTrackById, action.trackId);
  const { data, status } = response;
  if (status === 200) {
    yield put(successGetSingleTrackDetails(data));
  } else {
    yield put(failureGetTrackDetails(data));
  }
}
// Individual exports for testing
export default function* trackSaga() {
  yield takeLatest(REQUEST_GET_TRACK_DETAILS, getTrackDetails);
  yield takeLatest(REQUEST_GET_TRACK_DETAILS_BY_ID, getTrackDetailsById);
}
