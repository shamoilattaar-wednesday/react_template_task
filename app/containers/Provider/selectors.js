import { createSelector } from 'reselect';
import get from 'lodash/get';
import { initialState } from './reducer';

/**
 * Direct selector to the homeContainer state domain
 */

export const selectTrackDomain = (state) => state.track || initialState;

/**
 * Other specific selectors
 */

/**
 * Default selector used by HomeContainer
 */

export const selecttracksData = () => createSelector(selectTrackDomain, (substate) => get(substate, 'tracksData'));

export const selecttracksError = () => createSelector(selectTrackDomain, (substate) => get(substate, 'tracksError'));

export const selecttrackName = () => createSelector(selectTrackDomain, (substate) => get(substate, 'trackName'));

export const selectsingletrackData = () =>
  createSelector(selectTrackDomain, (substate) => get(substate, 'singletrackData'));

export const selecttrackId = () => createSelector(selectTrackDomain, (substate) => get(substate, 'trackId'));
