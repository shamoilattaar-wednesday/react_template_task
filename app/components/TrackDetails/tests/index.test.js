import React from 'react';
import { renderWithIntl, timeout } from '@utils/testUtils';
import { TrackDetailsTest as TrackDetails, mapDispatchToProps } from '../index';
import { fireEvent } from '@testing-library/dom';
import { trackTypes } from '@containers/Provider/reducer';

jest.mock('react-router-dom', () => ({
  useLocation: () => ({
    pathname: '/tracks/123456789',
    state: {
      item: {
        artworkUrl100: 'https://via.placeholder.com/100',
        collectionName: 'Test Collection',
        artistName: 'Test Artist',
        shortdescription: 'Test Description',
        collectionPrice: 9.99,
        previewUrl:
          'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/4a/68/06/4a6806aa-be22-3eff-8153-d9b1ab9dab8b/mzaf_12411707066990307220.plus.aac.p.m4a'
      }
    }
  })
}));

describe('<TrackDetails />', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<TrackDetails dispatchTrackDetails={submitSpy} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackDetails component', () => {
    const { getAllByTestId } = renderWithIntl(<TrackDetails dispatchTrackDetails={submitSpy} />);
    expect(getAllByTestId('track-card').length).toBe(1);
  });
  it('should render the track details inside the card', () => {
    let item = {
      trackId: '123456789',
      artworkUrl100: 'https://via.placeholder.com/100',
      collectionName: 'Test Collection',
      artistName: 'Test Artist',
      shortdescription: 'Test Description',
      collectionPrice: 9.99,
      previewUrl:
        'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/4a/68/06/4a6806aa-be22-3eff-8153-d9b1ab9dab8b/mzaf_12411707066990307220.plus.aac.p.m4a'
    };
    let items = { results: [item] };
    const { baseElement } = renderWithIntl(
      <TrackDetails dispatchTrackDetails={submitSpy} singletrackData={item} tracksData={items} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchTrackSearchSpy = jest.fn();
    const trackId = '123456789';
    const actions = {
      dispatchTrackDetails: { trackId, type: trackTypes.REQUEST_GET_TRACK_DETAILS_BY_ID },
      dispatchClearTrackDetails: { type: trackTypes.CLEAR_TRACK_DETAILS }
    };

    const props = mapDispatchToProps(dispatchTrackSearchSpy);
    props.dispatchTrackDetails(trackId);
    expect(dispatchTrackSearchSpy).toHaveBeenCalledWith(actions.dispatchTrackDetails);
  });
  it('should render the track details from the click on play and pause', async () => {
    let tracksData = {
      results: [
        {
          trackId: '123456789',
          artworkUrl100: 'https://via.placeholder.com/100',
          collectionName: 'Test Collection',
          artistName: 'Test Artist',
          shortDescription: 'Test Description',
          collectionPrice: 9.99,
          previewUrl:
            'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/4a/68/06/4a6806aa-be22-3eff-8153-d9b1ab9dab8b/mzaf_12411707066990307220.plus.aac.p.m4a'
        }
      ]
    };
    const { getByTestId } = renderWithIntl(
      <TrackDetails dispatchTrackDetails={submitSpy} tracksData={tracksData} singletrackData={tracksData} />
    );
    await timeout(500);

    fireEvent.click(getByTestId('play-audio'));
    await timeout(500);
    expect(getByTestId('pause-audio')).toBeInTheDocument();
    fireEvent.click(getByTestId('pause-audio'));
    await timeout(500);
    expect(getByTestId('play-audio')).toBeInTheDocument();
  });

  // sc
});
