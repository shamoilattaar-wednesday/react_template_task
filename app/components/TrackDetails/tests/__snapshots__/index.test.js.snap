// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackDetails /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root eh9f6zb0 css-e03web-MuiPaper-root-MuiCard-root"
      data-testid="track-card"
    >
      <div>
        <img
          src="https://via.placeholder.com/100"
        />
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="name"
        >
          Title name: Test Collection
        </p>
      </div>
      <p
        class="css-lh9t18 egtqi0h0"
        data-testid="artist-name"
      >
        Artist Name: Test Artist
      </p>
      <p
        class="css-lh9t18 egtqi0h0"
        data-testid="track-description"
      >
        Desciption: Test Description
      </p>
      <p
        class="css-lh9t18 egtqi0h0"
        data-testid="total-cost"
      >
        Total Cost: $9.99
      </p>
      <button>
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="play-audio"
        >
          Play Audio
        </p>
      </button>
      <audio
        class="audio-element"
      >
        <source
          src="https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/4a/68/06/4a6806aa-be22-3eff-8153-d9b1ab9dab8b/mzaf_12411707066990307220.plus.aac.p.m4a"
        />
      </audio>
    </div>
  </div>
</body>
`;

exports[`<TrackDetails /> should render the track details inside the card 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root eh9f6zb0 css-e03web-MuiPaper-root-MuiCard-root"
      data-testid="track-card"
    >
      <div>
        <img
          src="https://via.placeholder.com/100"
        />
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="name"
        >
          Title name: Test Collection
        </p>
      </div>
      <p
        class="css-lh9t18 egtqi0h0"
        data-testid="artist-name"
      >
        Artist Name: Test Artist
      </p>
      <p
        class="css-lh9t18 egtqi0h0"
        data-testid="track-description"
      >
        Desciption: Test Description
      </p>
      <p
        class="css-lh9t18 egtqi0h0"
        data-testid="total-cost"
      >
        Total Cost: $9.99
      </p>
      <button>
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="play-audio"
        >
          Play Audio
        </p>
      </button>
      <audio
        class="audio-element"
      >
        <source
          src="https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/4a/68/06/4a6806aa-be22-3eff-8153-d9b1ab9dab8b/mzaf_12411707066990307220.plus.aac.p.m4a"
        />
      </audio>
    </div>
  </div>
</body>
`;
